syntax = "proto3";

package gk.v1;

import "gk/v1/common.proto";
import "google/api/annotations.proto";
import "google/protobuf/struct.proto";

service Access {
	rpc Check(CheckAccessRequest) returns (CheckAccessResponse) {
		option (google.api.http) = {
			post : "/v1/check:access"
			body : "*"
		};
	}

	rpc CreatePolicy(CreateAccessPolicyRequest) returns (AccessPolicy) {
		option (google.api.http) = {
			post : "/v1/access-policies"
			body : "*"
		};
	}

	rpc RetrievePolicy(RetrieveAccessPolicyRequest) returns (AccessPolicy) {
		option (google.api.http) = {
			get : "/v1/access-policies/{id}"
		};
	}

	// Policy collections
	rpc AddPolicyCollection(AddAccessPolicyCollectionRequest) returns (AddAccessPolicyCollectionResponse) {
		option (google.api.http) = {
			post : "/v1/access-policies/{policy_id}/collections"
			body : "*"
		};
	}

	// Policy identities
	rpc AddPolicyIdentity(AddAccessPolicyIdentityRequest) returns (AddAccessPolicyIdentityResponse) {
		option (google.api.http) = {
			post : "/v1/access-policies/{policy_id}/identities"
			body : "*"
		};
	}
}

message AccessPolicy {
	string          id   = 1;
	optional string name = 2;

	repeated string           collection_ids = 3;
	repeated string           identity_ids   = 4;
	repeated AccessPolicyRule rules          = 5;
}

message AccessPolicyRule {
	google.protobuf.Struct attrs    = 2;
	string                 resource = 1;
}

message CheckAccessRequest {
	oneof identity {
		string identity_id  = 1;
		string identity_sub = 2;
	}

	string resource = 3;
}

message CheckAccessResponse {
	repeated Policy policies = 1;
}

message CreateAccessPolicyRequest {
	optional string id   = 1;
	optional string name = 2;

	repeated string           collection_ids = 3;
	repeated string           identity_ids   = 4;
	repeated AccessPolicyRule rules          = 5;
}

message RetrieveAccessPolicyRequest {
	string id = 1;
}

// Policy collections
message AddAccessPolicyCollectionRequest {
	string policy_id     = 1;
	string collection_id = 2;
}

message AddAccessPolicyCollectionResponse {}

// Policy identities
message AddAccessPolicyIdentityRequest {
	string policy_id   = 1;
	string identity_id = 2;
}

message AddAccessPolicyIdentityResponse {}
